DESIGN.md

```diff
@@ -1,31 +1,379 @@
+The main difference from the old CLI tool is command and output extensibility.
+
+Each command is defined in it's own module, implementing command behaviour. (see [Command behaviour](#command-behaviour))
+
+Output is processed with formatters and printers, which formats command output and renders it to somewhere. (see [Output formatting](#output-formatting))
+
+CLI core consists of several modules implementing command execution process:
+
+ - `RabbitMQCtl` - escript entry point. Command execution logic.
+ - `CommandModules` - responsible for command modules location
+ - `Helpers` - various functions to use in commands and main module
+ - `Config` - enviroment variable/argument values merge
+ - `Output` - output formatting logic
+ - `Parser` - OptionParser configuration
+
+### Command execution process:
+
+1. Arguments parsing
+
+ Command line arguments are parsed with OptionParser [http://elixir-lang.org/docs/stable/elixir/OptionParser.html]
+ Parser return a list of unnamed arguments and a map of options (named arguemtns)
+ First unnamed argument is a command name.
+ Named arguments can be global or command specific.
+ Command specific argument names and types are apecified in `switches/0` callback.
+ Global argument names are described in [Global arguments]
+
+2. Command discovery
+
+ If arguments list is not empty, its first element is considered a command name.
+ Command name is converted to CamelCase and a module with
+ `RabbitMQ.CLI.*.Commands.<CommandName>Command` name is selected as a command module.
+
+ List of available command modules depend on current tool scope (see [Command scopes](#command-scopes))
+
+3. Defaults and validation
+
+ After command module is found, effective command arguments are calculated by
+ merging global defaults and command specific defaults for both unnamed and named arguments.
+
+ Command specify defaults using `merge_defaults/2` callback (see [Command behaviour](#command-behaviour))
+
+ Arguments then validated using `validate/2` callback
+
+4. Command execution
+
+ Command is executed with `run/2` callback, which contains main command logic.
+ This callback can return any value.
+
+ Output from `run/2` callback is being processed with `output/2` callback,
+ which should format the output to specific type. (see [Output formatting](#output-formatting))
+
+ Output callback can specify return code to be returned by command
+
+5. Printing and formatting
+
+ `output/2` callback return value is being processed in `output.ex` module by
+ formatter and printer.
+
+ Formatter translates the output value to a sequence of strings or error value.
+ Example formatters are `json`, `csv` and `erlang`
+
+ Printer renders formatted strings to output device (e.g. stdout, file)
+
+6. Return
+
+ Errors during execution (e.g. validation failures, command errors) are being printed
+ to `stderr`
+
+ If command has failed to execute, non-zero code is returned.
```

Comment:
If command has failed to execute, non-zero code is returned => If a command has failed to execute, a non-zero code is returned